http/tus.go

```diff
@@ -0,0 +1,242 @@
+	if !ok {
+		// If we don't define an absolute URL for tusd, it creates an absolute URL for us that the client will use.
+		// See tusd/handler/unrouted_handler.go/absFileURL() for details.
+		// This URL's scheme will be http in our case (as we don't use tusd's inbuilt TLS feature),
+		// which is fine if we don't use both a browser and a reverse proxy that terminates SSL for us.
+		// In case we do, we need to define an absolute URL with the correct scheme, or we'll get mixed content errors.
+		// We can extract the correct scheme and host from the origin request header, if it exists (which always is the case for browsers).
+		var origin string
+		if originHeader, ok := r.Header["Origin"]; ok && len(originHeader) > 0 {
+			origin = originHeader[0]
+		}
+		basePath, err := url.JoinPath(origin, th.server.BaseURL, th.apiPath) 
+		if err != nil {
+			return nil, err
+		}
+
+		log.Printf("Creating tus handler for user %s on path %s\n", d.user.Username, basePath)
+		handler, err := th.createTusHandler(d, basePath)
+		if err != nil {
+			return nil, err
+		}
+		th.handlers[d.user.ID] = handler
+	}
+
+	return handler, nil
+}
+
+func (th tusHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	code, err := withUser(func(w http.ResponseWriter, r *http.Request, d *data) (int, error) {
+		// Create a new tus handler for current user if it doesn't exist yet
+		handler, err := th.getOrCreateTusHandler(d, r)
+
+		if err != nil {
+			return 400, err
+		}
+
+		// Create upload directory for each request
+		uploadDir := filepath.Join(d.user.FullPath("/"), ".tmp_upload")
+		if err := os.MkdirAll(uploadDir, os.ModePerm); err != nil {
+			return http.StatusInternalServerError, err
+		}
+
+		switch r.Method {
+		case "POST":
+			handler.PostFile(w, r)
+		case "HEAD":
+			handler.HeadFile(w, r)
+		case "PATCH":
+			handler.PatchFile(w, r)
+		default:
+			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
+		}
+
+		// Isn't used
+		return 201, nil
+	})(w, r, &data{
+		store:    th.store,
+		settings: th.settings,
+		server:   th.server,
+	})
+
+	switch {
+	case err != nil:
+		http.Error(w, err.Error(), code)
+	case code >= 400:
+		http.Error(w, "", code)
+	}
+}
+
+func (th tusHandler) createTusHandler(d *data, basePath string) (*tusd.UnroutedHandler, error) {
+	uploadDir := filepath.Join(d.user.FullPath("/"), th.uploadDirName)
+	tusStore := filestore.FileStore{
+		Path: uploadDir,
+	}
+	composer := tusd.NewStoreComposer()
+	tusStore.UseIn(composer)
+	if composer.UsesConcater && composer.UsesLocker {
+		log.Printf("WARNING: tusd's concater extension is enabled. This is not recommended as it can cause data loss. ")
+	}
+
+	handler, err := tusd.NewUnroutedHandler(tusd.Config{
+		BasePath:              basePath,
+		StoreComposer:         composer,
+		NotifyCompleteUploads: true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create handler: %s", err)
+	}
+
+	// Goroutine to handle completed uploads
+	go func() {
```

Comment:
I wouldn't say I like the idea of hanging forever goroutine per user. In my opinion, it should be implemented differently.